src/libappimage/libappimage.cpp

```diff
@@ -0,0 +1,364 @@
+#endif
+
+using namespace appimage::core;
+
+namespace bf = boost::filesystem;
+
+extern "C" {
+
+/* Check if a file is an AppImage. Returns the image type if it is, or -1 if it isn't */
+int appimage_get_type(const char* path, bool verbose) {
+    typedef std::underlying_type<AppImageFormat>::type utype;
+
+    try {
+        AppImage appImage(path);
+        return static_cast<utype>(appImage.getFormat());
+    } catch (const std::runtime_error& err) {
+        if (verbose)
+            std::clog << "Error at " << __FUNCTION__ << " : " << err.what() << std::endl;
+    } catch (...) {
+        if (verbose)
+            std::clog << "Error at " << __FUNCTION__ << " : that's all we know." << std::endl;
+    }
+
+    return static_cast<utype>(AppImageFormat::INVALID);
+}
+
+char** appimage_list_files(const char* path) {
+    char** result = nullptr;
+    try {
+        AppImage appImage(path);
+
+        std::vector<std::string> files;
+        for (auto itr = appImage.files(); itr != itr.end(); ++itr)
+            if (!(*itr).empty())
+                files.emplace_back(*itr);
+
+
+        result = static_cast<char**>(malloc(sizeof(char*) * (files.size() + 1)));
+        for (int i = 0; i < files.size(); i++)
+            result[i] = strdup(files[i].c_str());
+
+        result[files.size()] = nullptr;
+
+        return result;
+    } catch (const std::runtime_error& err) {
+        std::clog << "Error at " << __FUNCTION__ << " : " << err.what() << std::endl;
+    } catch (...) {
+        std::clog << "Error at " << __FUNCTION__ << " : that's all we know." << std::endl;
+    }
+
+    // Create empty string list
+    result = static_cast<char**>(malloc(sizeof(char*)));
+    result[0] = nullptr;
+
+    return result;
+}
+
+void appimage_string_list_free(char** list) {
+    for (char** ptr = list; ptr != NULL && *ptr != NULL; ptr++)
+        free(*ptr);
+
+    free(list);
+}
+
+bool
+appimage_read_file_into_buffer_following_symlinks(const char* appimage_file_path, const char* file_path, char** buffer,
+                                                  unsigned long* buf_size) {
+    // Init output params
+    *buffer = nullptr;
+    *buf_size = 0;
+
+    std::string targetFile = file_path;
+    std::vector<std::string> visitedEntries;
+
+    while (!targetFile.empty()) {
+        // Find loops
+        if (std::find(visitedEntries.begin(), visitedEntries.end(), targetFile) != visitedEntries.end()) {
+            std::clog << "WARNING: Links loop found while extracting " << file_path;
+            break;
+        }
+
+        visitedEntries.emplace_back(targetFile);
+        std::string nextHoop;
+        try {
+            AppImage appImage(appimage_file_path);
+
+            for (auto itr = appImage.files(); itr != itr.end(); ++itr) {
+                if (itr.path() == targetFile) {
+                    if (itr.type() == PayloadEntryType::REGULAR) {
+                        auto data = std::vector<char>(std::istreambuf_iterator<char>(itr.read()),
```

Comment:
Isn't there an easier way to populate that vector? It's quite lengthy, and also really inefficient due to the vast amount of allocations and memory I/O whenever the vector must be copied. Can't we get the file size in advance, create a new vector with the right size, and then use `std::copy` to populate the vector?

This is used more than once in this file, and should therefore go into a helper function.

Reply:
IFAIK this is the shortest way.
Getting the file size in advance is not supported yet, haven't investigated if it's supported by `libarchive` and `squashfuse`.

About creating a helper, this is a single line instruction. There will be no much difference.

Reply:
I wrote elsewhere that these `read()` should probably then implement this algorithm to avoid redundancy. We can provide access to the stream buffer with a separate getter. That also allows for implementing a better option once we know how to guess the filesize (it can be done type specific then, even!).